library/core/src/iter/adapters/map_windows.rs

```diff
@@ -0,0 +1,379 @@
+use crate::{
+    fmt,
+    iter::{ExactSizeIterator, FusedIterator, TrustedLen},
+    mem::{self, MaybeUninit},
+    ptr,
+};
+
+/// An iterator over the mapped windows of another iterator.
+///
+/// This `struct` is created by the [`Iterator::map_windows`]. See its
+/// documentation for more information.
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+#[unstable(feature = "iter_map_windows", reason = "recently added", issue = "87155")]
+pub struct MapWindows<I: Iterator, F, const N: usize> {
+    f: F,
+    inner: MapWindowsInner<I, N>,
+}
+
+struct MapWindowsInner<I: Iterator, const N: usize> {
+    iter: I,
+    // Since iterators are assumed lazy, i.e. it only yields an item when
+    // `Iterator::next()` is called, and `MapWindows` is not an exception.
+    //
+    // At the first iteration, we need to collect `N` elements from
+    // `MapWindows::iter` (calling it the item iterator for convenient) into
+    // the buffer, but before that, it is uninitialized and thus `buffer` is
+    // `None`.
+    initialized: bool,
```

Comment:
I think we should find a different name for this field, because whether the buffer is initialized or not can be tracked directly by the `Option`. Seeing a separate field for that makes me confused and it's not clear what exactly we're tracking here.

In practice I think the only reason we actually need a separate field is to track when the underlying iterator is *done*, so that we can avoid calling next on it repeatedly if there's Nones already interspersed. It seems like we could call this field `is_done` or store the underlying iterator in an Option as well? (I think wrapping the underlying iterator in Fuse could also work, but is trickier to implement since there's no is_done method on fused).

Regardless, we should update the comment on this field to reflect whatever we come up with.

Reply:
To clarify, I list the three states of the iterator:
1. Uninitialized (`initialized == false && buffer.is_none()` in current code), before any `.next()` or `.advance_by()` called on the `MapWindows` iterator;
2. Initialized and active (`initialized == true && buffer.is_some()` in current code), since `.next()` or `.advance_by` called but not fused yet;
3. Fused (`initialized == true && buffer.is_none()` in current code), since the inner iterator yields a `None` and fused by us.

Your proposal is to distinguish states 1 & 3 with `is_done`, making it work similarly like `Fuse`. I think we may not need `is_done` if we wrap the `iter: I` with `iter: Fuse<I>` because everywhere we need `is_done` is to call `.next()`, `.advance_by()`, or `.size_hint()` etc. which is already wrapped in `Fuse`.

Reply:
I'd like to add an `is_done` to `Fuse` and that will help to track this state here.

Reply:
Oh, I realize what you have said. You meant that `.next()` would invoke the mapper function `f`, but `advance_by` would skip those mapping. They are unequivalent, is it?